sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerAsyncClient.java

```diff
@@ -0,0 +1,326 @@
+    }
+
+    /**
+     * Enqueue a single event with send options.
+     *
+     * @param eventData The {@link EventData} will be stored in client temporarily and sent to event hub.
+     * @param options A set of options used to configure the {@link EventDataBatch}.
+     * @throws AmqpException if the size of {@code events} exceed the maximum size of a single batch.
+     * @return Gets a {@link Mono} that completes when eventData enqueued.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> enqueueEvent(EventData eventData, SendOptions options) {
+        List<EventData> list = new ArrayList<>();
+        list.add(eventData);
+        return enqueueEvents(list, options);
+    }
+
+    /**
+     * Enqueue iterable events without send options, will be sent to event hub by default send options.
+     *
+     * @param events Events will be stored in client temporarily and sent to event hub.
+     * @throws AmqpException if the size of {@code events} exceed the maximum size of a single batch.
+     * @return Gets a {@link Mono} that completes when eventData enqueued.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> enqueueEvents(Iterable<EventData> events) {
+        return enqueueEvents(events, DEFAULT_SEND_OPTIONS);
+    }
+
+    /**
+     * Enqueue iterable events with send options.
+     *
+     * @param events Events will be stored in client temporarily and sent to event hub.
+     * @param options A set of options used to configure the {@link EventDataBatch}.
+     * @throws AmqpException if the size of {@code events} exceed the maximum size of a single batch.
+     * @return Gets a {@link Mono} that completes when eventData enqueued.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> enqueueEvents(Iterable<EventData> events, SendOptions options) {
+        Objects.requireNonNull(events, "'eventData' cannot be null.");
+        CreateBatchOptions createBatchOptions = new CreateBatchOptions().setPartitionId(options.getPartitionId()).setPartitionKey(options.getPartitionKey());
+        AtomicReference<EventDataBatch> currentBatch = new AtomicReference<>();
+        ConcurrentLinkedQueue<EventData> failEvents = new ConcurrentLinkedQueue<>();
+
+        Flux<EventData> data = Flux.fromIterable(events);
+        return Mono.when(
+            data.flatMap(eventData -> {
+                EventDataBatch currentEventBatch = currentBatch.get();
+                if (Objects.isNull(currentEventBatch) || !currentEventBatch.tryAdd(eventData)) {
+                    if (!Objects.isNull(currentBatch.get())) {
+                        EventDataBatch fullBatch = currentBatch.getAndSet(null);
+                        addBatchToPartitionMap(options.getPartitionId(), fullBatch);
+                    }
+                    return producer.createBatch(createBatchOptions).map(eventDataBatch -> {
+                        currentBatch.set(eventDataBatch);
```

Comment:
Why `set` again? What is the purpose of this `currentBatch`? This part is confusing. Either add comments about the flow, or thing about whether this is just wrong code.

Reply:
The purpose of `currentBatch` is used for a scenario that we add a event to a batch, it reach the max size of a batch, so I need to add this current batch to batch deque, and create a new batch as current batch to add the rest of events.

Reply:
The problem is whether you code really works correctly in parallels.

E.g. if thread1 and thread2 of `flatMap` get to this line at same time (I assume `currentBatch` is emtpy at start), which one is the `currentBatch` (at least you waste a not-full batch)?

Reply:
OK, I know what you mean. Actually I need a synchronous lock method to get current batch that if current batch is full, I need to add current batch to the queue and create a new batch replace it.

Reply:
You likely need to rethink the approach.

If you do lock in async method, the code is likely wrong in the first place...